src/viztracer/modules/vcompressor/vc_dump.c

```diff
@@ -146,6 +147,57 @@ int dump_parsed_trace_events(PyObject* trace_events, FILE* fptr)
         }
     }
 
+    // Iterate through instant events
+    ppos = 0;
+    PyObject* dumps_func = NULL;
+    PyObject* dumps_func_args = PyTuple_New(1);
+    PyObject* args_object = NULL;
+    dumps_func = PyObject_GetAttrString(json_module, "dumps");
+    if (!dumps_func) {
+        goto clean_exit;
+    }
+
+    while (PyDict_Next(instant_events, &ppos, &key, &value)) {
+        // head
+        fputc(VC_HEADER_INSTANT_EVENT, fptr);
+
+        // pid-tid-name-scope
+        uint64_t pid = PyLong_AsLong(PyTuple_GetItem(key, 0));
+        uint64_t tid = PyLong_AsLong(PyTuple_GetItem(key, 1));
+        const char* name = PyUnicode_AsUTF8(PyTuple_GetItem(key, 2));
+        const char* scope = PyUnicode_AsUTF8(PyTuple_GetItem(key, 3));
+        fwrite(&pid, sizeof(uint64_t), 1, fptr);
+        fwrite(&tid, sizeof(uint64_t), 1, fptr);
+        fwritestr(name, fptr);
+        fwritestr(scope, fptr);
+
+        // count
+        uint64_t ts_size = PyList_GET_SIZE(value);
+        fwrite(&ts_size, sizeof(uint64_t), 1, fptr);
+
+        for (Py_ssize_t idx = 0; idx < (Py_ssize_t)ts_size / 2; idx++) {
+            // ts
+            double ts = PyFloat_AsDouble(PyList_GET_ITEM(value, 2 * idx));
+            int64_t ts64 = ts * 1000;
+            fwrite(&ts64, sizeof(int64_t), 1, fptr);
+
+            // args
+            PyTuple_SetItem(dumps_func_args, 0, PyList_GET_ITEM(value, 2 * idx + 1));
+            args_object = PyObject_CallObject(dumps_func, dumps_func_args);
+            const char* args = PyUnicode_AsUTF8(args_object);
+            fwritestr(args, fptr);
```

Comment:
I believe there's some reference bugs here. Remember that this piece of code is in a loop. For `dump_func_args`, `PyTuple_SetItem` will steal a reference that's borrowed from `value`. Because you are reusing `dump_func_args`, the previous item on the tuple will be discarded(the reference will be decreased). So we lost a reference on all the elements in `value`.

If you do want to reuse the `dump_func_args` tuple, you should probably increase the reference to all the element assigned to it. 

In a similar way, `args_object` is replaced without properly released in each loop, which leaves all the results for the function call hanging there forever.

In another matter, `args` should be defined on the top of this code block. Or you could simply skip that variable because there's no reference involved here.